### ISSUE TITLE
Next steps

### ISSUE THREAD

**Issue by jlesquembre** [jlesquembre/clj-nix#10]
I'm summarizing the next steps for the project. Suggestions are welcome :)

- [x] Rethink (again) the lock format. After clj-nix first release, [clojure-nix-locker](https://github.com/bevuta/clojure-nix-locker) showed that it's possible to create a maven and gitlibs cache to avoid network requests at build time. The advantage is that we can compute the classpath with Clojure at build time. While I was able to accomplish the same in the `next` branch (without a cache), the [clojure-nix-locker](https://github.com/bevuta/clojure-nix-locker) approach it's simpler
- [x] Create only one lock file per project. There are projects with multiple deps.edn files. In `clj-nix` current version, we need to generate a lock file per deps.edn file, I want to be able to generate only one lock file for the whole project
- [x] Improve lock file generation. The current approach is quite naive, it downloads every time all the maven/git dependencies. It would be nice to reuse the system cache, to avoid network issues (see #9) and make the lock generation faster.
- [x] Allow executing a random `build.clj` at build time, while providing a default one. Currently, the way we are generating a jar for the project cannot be modified. I think that in most cases the default way of generating the jar should be enough, but I want to add an option in case you want to use your custom `build.clj`
- [x] Add a new function to build a library jar (see #4). This one should be easy, we are already generating a jar anyways 

In short, the goal is to be able to integrate `clj-nix` in any clojure project without requiring any changes on the project (other than adding a flake.nix file)

**Comment by PlumpMath** [comment]
I have a question about the lock file.
Once you're done with what you're doing,
Does it automatically generate automatically after receiving the src instead of manually generating it?

**Comment by jlesquembre** [comment]
@PlumpMath you have to generate a new lock file manually after I finish with the new lock file format.
The same applies if you change your dependencies in your deps.edn, which could be automated with github actions or any other CI/CD tool

**Comment by jaen** [comment]
Re: lock format, maybe you could consider using https://github.com/nix-community/dream2nix's lock format?

While dream2nix doesn't yet provide external extensibility (so each language ecosystem supported has to be inside the repo), I do plan to implement I for my ends, and then it would be easier to add dream2nix support to clj-nix, if the lock format is already the same. As a bonus, it already supports Node, so it would take care of Node dependencies for Cljs.
